pkg/endpoint/bpf.go

```diff
@@ -1174,6 +1227,59 @@ func (e *Endpoint) applyPolicyMapChanges() error {
 	//  applied to the Endpoint's bpf policy map.
 	closer, changes := e.desiredPolicy.ConsumeMapChanges()
 	defer closer()
+	var deleteLen int
+	for k := range changes.Deletes {
+		if _, ok := changes.Adds[k]; !ok {
+			deleteLen++
+		}
+	}
+	// If the desiredPolicy will be larger than the BPF maximum after
+	// the changes we need to reset the consumed changes and either
+	// lockdown or return an inevitable error (which would happen on
+	// insertion if we proceed).
+	if e.desiredPolicy.GetPolicyMap().Len()+len(changes.Adds)-deleteLen > policymap.MaxEntries {
+		// Revert the map changes that were just consumed.
+		e.desiredPolicy.RevertChanges(changes)
+
+		// Avoid an insertion error further down and return an
+		// error now.
+		if !option.Config.EnableEndpointLockdownOnPolicyOverflow {
+			// Do not return ErrPolicyEntryMaxExceeded as it
+			// implies that option.Config.EnableEndpointLockdownOnPolicyOverflow
+			// is "true".
+			return fmt.Errorf("policy map exceeds the map max entries limit")
+		}
+
+		e.getLogger().WithFields(logrus.Fields{
+			logfields.EndpointID: e.ID,
+		}).Warnf("The policy map exceeds the max entries limit, %s is enabled, locking the endpoint down.",
+			option.EnableEndpointLockdownOnPolicyOverflow)
+
+		if err := e.endpointPolicyLockdown(); err != nil {
```

Comment:
Does a call to this function need to happen if the endpoint is already locked? If not we should wrap it in an `if !e.lockdown {}` block or `endpointPolicyLockdown()` could return early `if e.lockdown == true`. Otherwise, IIUC, we keep recreating the map to insert the deny all entries.

Reply:
> Does a call to this function need to happen if the endpoint is already locked?

Yes it does.

> Otherwise, IIUC, we keep recreating the map to insert the deny all entries.

Correct.

Reply:
Could you elaborate why is it necessary to do this if the endpoint is already locked down? Is it to accommodate agent restart and upgrade scenarios?